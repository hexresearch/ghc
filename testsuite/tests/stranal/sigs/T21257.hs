module T21257 (g, g2) where

f :: (Bool, Bool) -> (Bool, Bool)
f pr = (pr `seq` True, case pr of (a,b) -> a && b)
{-# NOINLINE f #-}

-- The demand signature of `g` should say `LP(ML,ML)`, e.g., that the pair
-- components are used at most once.
g :: (Bool, Bool) -> ()
g pr = f pr `seq` ()

-- | Key point: Unlike for `seq`, repeated apps of `seq'` can't be simplified
-- away.
seq' :: a -> b -> b
seq' = seq
{-# NOINLINE seq' #-}

f2 :: (Bool, Bool) -> ()
f2 pr = pr `seq'` fst pr `seq'` ()
{-# NOINLINE f2 #-}

-- The demand signature of `g2` should say `SP(1L,A)` (prior to worker/wrapper),
-- e.g., that the pair components are used at most once.
g2 :: (Bool, Bool) -> ()
g2 pr = f2 pr `seq` ()
