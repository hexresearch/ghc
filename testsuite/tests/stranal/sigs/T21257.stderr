
==================== Strictness signatures ====================
T21257.$trModule:
T21257.g: <LP(ML,ML)>
T21257.g2: <SP(1L,A)>



==================== Cpr signatures ====================
T21257.$trModule:
T21257.g: 1
T21257.g2:



==================== Strictness signatures ====================
T21257.$trModule:
T21257.g: <LP(ML,ML)>
T21257.g2: <1P(1L,A)>


